### PR TITLE
[feat][FE] Stop persisting new empty records

### LINKED ISSUE #4773: On RecordTable, I can create an empty record without persisting

### PR DESCRIPTION
**Pull request by ijreilly**
## Context
Closes [#4773](https://github.com/twentyhq/twenty/issues/4773)
Persisting of new records is delayed to cell escape and not performed for empty records.

## How was it tested?
Locally tested + jest

### CODE REVIEW

File: packages/twenty-front/src/pages/object-record/RecordIndexPage.tsx [hunk 1]
```diff
@@ -20,27 +20,19 @@ const StyledIndexContainer = styled.div`
...
 
   const { setSelectedTableCellEditMode } = useSelectedTableCellEditMode({
     scopeId: recordIndexId,
   });
 
+  const { setPendingRecordId } = usePendingRecordId();
+
   const handleAddButtonClick = async () => {
-    await createOneObject?.({
-      position: 'first',
```
Comment (ijreilly) [inline packages/twenty-front/src/pages/object-record/RecordIndexPage.tsx:40]: not sure what that is for, did not use it afterwards

File: packages/twenty-front/src/modules/object-record/record-field/hooks/usePersistField.ts [hunk 2]
```diff
@@ -31,15 +36,46 @@ import { isFieldRelationValue } from '../types/guards/isFieldRelationValue';
...
-    fieldDefinition,
-    useUpdateRecord = () => [],
-  } = useContext(FieldContext);
+const useUpsertRecord = () => {
+  const { pendingRecordId } = usePendingRecordId();
 
+  const { useUpdateRecord = () => [] } = useContext(FieldContext);
   const [updateRecord] = useUpdateRecord();
 
+  const objectNamePlural = useParams().objectNamePlural ?? '';
```
Comment (ijreilly) [inline packages/twenty-front/src/modules/object-record/record-field/hooks/usePersistField.ts:45]: is it safe to call this here ? or should usePersistField remain agnostic to current page params?

Comment (charlesBochet) [reply packages/twenty-front/src/modules/object-record/record-field/hooks/usePersistField.ts:45]: discussed with you, we can leverage fieldContext!

File: packages/twenty-front/src/modules/object-record/record-field/hooks/usePersistField.ts [hunk 3]
```diff
@@ -31,15 +36,46 @@ import { isFieldRelationValue } from '../types/guards/isFieldRelationValue';
...
 import { isFieldTextValue } from '../types/guards/isFieldTextValue';
 
-export const usePersistField = () => {
-  const {
-    entityId,
-    fieldDefinition,
-    useUpdateRecord = () => [],
-  } = useContext(FieldContext);
+const useUpsertRecord = () => {
+  const { pendingRecordId } = usePendingRecordId();
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-field/hooks/usePersistField.ts:40]: I am not sure this upsert logic should be within PersistField. It does not seem to be the responsability of the field to know about the table row. I would put in in the RecordTableCell

File: packages/twenty-front/src/modules/object-record/record-table/hooks/usePendingRecordId.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { useRecoilState } from 'recoil';
+
+import { pendingRecordIdState } from '@/object-record/record-table/states/pendingRecordIdState';
+
+export const usePendingRecordId = () => {
+  const [pendingRecordId, setPendingRecordId] =
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-table/hooks/usePendingRecordId.ts:6]: I don't think this is a good pattern to introduce a hook to wrap useRecoilState.
The issue is that this hook will subscribe to the state (and trigger re-renders) even if you are only interested in the setter. I would call directly useRecoilState, useRecoilValue, useSetRecoilState where needed in your code

File: packages/twenty-front/src/modules/object-record/record-table/record-table-cell/components/RecordTableCell.tsx [hunk 5]
```diff
@@ -17,17 +20,27 @@ export const RecordTableCell = ({
...
+  const draftValue = useRecoilValue(getDraftValueSelector());
+
+  const shouldPersistField =
+    (pendingRecordId && draftValue) || !pendingRecordId;
+
   const { moveLeft, moveRight, moveDown } = useRecordTableMoveFocus();
 
   const handleEnter: FieldInputEvent = (persistField) => {
-    persistField();
+    if (shouldPersistField) persistField();
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-table/record-table-cell/components/RecordTableCell.tsx:36]: Let's add the logic within usePersistField here actually :)

File: packages/twenty-front/src/modules/object-record/record-table/states/pendingRecordIdState.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { atom } from 'recoil';
+
+export const pendingRecordIdState = atom<string | null>({
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-table/states/pendingRecordIdState.ts:3]: use createState

File: packages/twenty-front/src/pages/object-record/RecordIndexPage.tsx (hunk 1 above)
Comment (charlesBochet) [reply packages/twenty-front/src/pages/object-record/RecordIndexPage.tsx:40]: it's to add the record to the top in the case of an un-ordered list (list are always ordered by position after all other sorts), we need it!

**Review by charlesBochet — APPROVED** [review]
Looks great!

**Pull request merged**